dwave/optimization/include/dwave-optimization/array.hpp

```diff
@@ -237,338 +237,473 @@ struct Update {
-                loc[i] = 0;
+        ArrayIteratorImpl_& operator--() {
+            if (shape_ && mask_) {
+                assert(false && "not implemented yet");  // or maybe ever
+                unreachable();
+            } else if (shape_) {
+                ptr_ += shape_->decrement1() / itemsize;
+            } else if (mask_) {
+                // decrement both the mask ptr and the data ptr
+                --(mask_->mask_ptr);
+                --ptr_;
+            } else {
+                --ptr_;
             }
+            return *this;
         }
-
-        ShapeInfo(const ShapeInfo& other) noexcept
-                : ShapeInfo(other.ndim, other.shape, other.strides) {
-            if (ndim >= 1) std::copy(other.loc.get(), other.loc.get() + ndim - 1, loc.get());
+        ArrayIteratorImpl_ operator--(int) {
+            ArrayIteratorImpl_ tmp = *this;
+            --(*this);
+            return tmp;
         }
 
-        ShapeInfo(ShapeInfo&& other) = default;
-
-        // Both the copy and move operator using the copy-and-swap idiom
-        ShapeInfo& operator=(ShapeInfo other) noexcept {
-            using std::swap;  // ADL, if it matters
-            std::swap(ndim, other.ndim);
-            std::swap(shape, other.shape);
-            std::swap(strides, other.strides);
-            std::swap(loc, other.loc);
+        ArrayIteratorImpl_& operator+=(difference_type rhs) {
+            if (shape_ && mask_) {
+                assert(false && "not implemented yet");  // or maybe ever
+                unreachable();
+            } else if (shape_) {
+                ptr_ += shape_->increment(rhs) / itemsize;
+            } else if (mask_) {
+                ptr_ += rhs;
+                (mask_->mask_ptr) += rhs;
+            } else {
+                ptr_ += rhs;
+            }
             return *this;
         }
+        ArrayIteratorImpl_& operator-=(difference_type rhs) { return this->operator+=(-rhs); }
 
-        // returns the number of bytes to advance. Note that this is in bytes!
-        std::ptrdiff_t advance() {
-            std::ptrdiff_t offset = 0;
-
-            ssize_t dim = ndim - 1;
-            for (; dim >= 1; --dim) {
-                offset += strides[dim];
-
-                // we don't need to advance any other dimensions
-                if (++loc[dim - 1] < shape[dim]) break;
+        friend ArrayIteratorImpl_ operator+(ArrayIteratorImpl_ lhs, difference_type rhs) {
+            return lhs += rhs;
+        }
+        friend ArrayIteratorImpl_ operator+(difference_type lhs, ArrayIteratorImpl_ rhs) {
+            return rhs += lhs;
+        }
+        friend ArrayIteratorImpl_ operator-(ArrayIteratorImpl_ lhs, difference_type rhs) {
+            return lhs -= rhs;
+        }
 
-                offset -= strides[dim] * shape[dim];  // zero out this dim
-                loc[dim - 1] = 0;
+        friend difference_type operator-(const ArrayIteratorImpl_& a, const ArrayIteratorImpl_& b) {
```

Comment:
I contemplated moving this (and a few other similar methods) into the `.cpp` file. But to my great surprise the compiler claims that it did indeed get inlined. So because it's performance-critical I decided to keep it in the header. Probably doesn't actually matter but :shrug: .